### PR TITLE
BWA/STAR dynamic index check

### LINKED ISSUE #27: Make check for BWA/STAR index dynamic
**Is your feature request related to a problem? Please describe.**
Currently, the ngs_mapping/bwa/path_index attribute always needs to be filled in in the config.yaml, even if the project only includes RNA libs. And similarly, as of some recent updates the star index must be specified even if the project only includes DNA libs.

**Describe the solution you'd like**
_The purpose of this ticket is to make the code smarter, and only check if the index path exists when applicable._

Details of the lib type(s) included in the project will come from the **biomedsheets tsv file**.  
(In the interest of not overly complicating things we'll use the entire contents of the sheet, regardless of whether we may only be actively processing a subset of the entries)

**Describe alternatives you've considered**
N/A

**Additional context**
Internal ticket \#79.

### PR DESCRIPTION
**Pull request by eudesbarbosa**
closes #27 
addresses #6 

**Changes**

* Emptied `DEFAULT_CONFIG` aligners in NGS Mapping Wokrflow. User must define in local config file.
* Fixed a couple of Codacy minor issues.
* Reformatted and documented the following methods:
   -  `BaseStep::ensure_w_config()`;
   - All `check_config()` in the NGS mapping worflow file.

### CODE REVIEW

File: tests/snappy_pipeline/workflows/conftest.py [hunk 1]
```diff
@@ -151,7 +158,8 @@ def sample_cache_dict():
 def germline_sheet_fake_fs(fake_fs, germline_sheet_tsv):
     """Return fake file system setup with files for the germline_sheet_tsv"""
     # Create work directory
-    fake_fs.fs.create_dir("/work")
```
Comment (holtgrewe) [inline tests/snappy_pipeline/workflows/conftest.py:154]: rather use `exists_ok=True`

File: snappy_pipeline/workflows/ngs_mapping/__init__.py [hunk 2]
```diff
@@ -587,33 +607,44 @@ class NgmlrStepPart(ReadMappingStepPart):
...
 class ExternalStepPart(ReadMappingStepPart):
     """Support for linking in external BAM files"""
 
     name = "external"
 
+    def update_cluster_config(self, cluster_config):
+        cluster_config["ngs_mapping_external_run"] = {"mem": 1024, "time": "04:00", "ntasks": 1}
+
     def check_config(self):
+        """Method checks that all parameters required to execute BWA are present in the
```
Comment (holtgrewe) [inline snappy_pipeline/workflows/ngs_mapping/__init__.py:641]: I think we should wrap also comments to 100 characters.

Also I like to stick to having a one sentence/line summary first, then an empty line, then the rest of the docs.

File: tests/snappy_pipeline/workflows/conftest.py (hunk 1 above)
Comment (holtgrewe) [reply tests/snappy_pipeline/workflows/conftest.py:154]: same elsewhere

File: tests/snappy_pipeline/workflows/test_workflows_gene_expression_quantification.py [hunk 3]
```diff
@@ -26,6 +26,14 @@ def minimal_config():
           reference:
             path: /path/to/ref.fa
```
Comment (holtgrewe) [inline tests/snappy_pipeline/workflows/test_workflows_gene_expression_quantification.py:28]: Please stick to two space indentation for YAML.

File: tests/snappy_pipeline/workflows/test_workflows_somatic_msi_calling.py [hunk 4]
```diff
@@ -109,4 +120,8 @@ def test_somatic_msi_calling_workflow(somatic_msi_calling_workflow):
         "output/mantis.bwa.P002-T2-DNA1-WGS1/out/mantis.bwa.P002-T2-DNA1-WGS1_results.txt",
         "output/mantis.bwa.P002-T2-DNA1-WGS1/out/mantis.bwa.P002-T2-DNA1-WGS1_results.txt.status",
     ]
-    assert set(somatic_msi_calling_workflow.get_result_files()) == set(expected)
+    observed = set(somatic_msi_calling_workflow.get_result_files())
+    expected = set(expected)
+    print(observed)
```
Comment (holtgrewe) [inline tests/snappy_pipeline/workflows/test_workflows_somatic_msi_calling.py:125]: These should probably go away.

**Review by holtgrewe — COMMENTED** [review]
Overall, looks good and I'm seeing mostly whitespace issues.

File: snappy_pipeline/workflows/ngs_mapping/__init__.py (hunk 2 above)
Comment (eudesbarbosa) [reply snappy_pipeline/workflows/ngs_mapping/__init__.py]: Reviewed. Please check if that is what you had in mind.

File: tests/snappy_pipeline/workflows/test_workflows_gene_expression_quantification.py (hunk 3 above)
Comment (eudesbarbosa) [reply tests/snappy_pipeline/workflows/test_workflows_gene_expression_quantification.py:28]: Fixed.

File: tests/snappy_pipeline/workflows/test_workflows_somatic_msi_calling.py (hunk 4 above)
Comment (eudesbarbosa) [reply tests/snappy_pipeline/workflows/test_workflows_somatic_msi_calling.py]: Fixed.

**Comment by eudesbarbosa** [comment]
@holtgrewe, I will **keep** this as a **draft** for now.

Although it fix the problem of no longer being necessary to set parameters that are won't be used, it is more restricted to the config structure. That might be a hard pill to swallow for users. Defining the tools used in ngs mapping is mandatory now, example:

```
ngs_mapping:
  tools:
    dna: ['bwa']
```

**Comment by holtgrewe** [comment]
To clarify, the only difference is that people will have to touch the ngs_mapping configuration and add bwa as a mapper? That's fine I guess.

**Comment by eudesbarbosa** [comment]
Yes, basically add the following lines in the config, then the free from defining the other required.

```
step_config:
  ngs_mapping:
    tools:
      dna: ['bwa']
```

**Pull request merged**